### PR TITLE
irmin: use an optimized short hash function for crypto hashes

### LINKED ISSUE #1164: layers benchmarks are failing with "key not found"
The easiest way to repro:

```
$ dune exec -- ./bench/irmin-pack/layers.exe -n 3000000
[... wait ~30s ...]
Benchmarks for layered store: internal error, uncaught exception:
                              (Failure "a65756b629e51a22451d93a5e4a05fef6685d5c0: unknown key")
```

### PR DESCRIPTION
**Pull request by pascutto**
Reintroduces #1143, and fixes #1164.

### CODE REVIEW

File: src/irmin-pack/pack_index.ml [hunk 1]
```diff
@@ -35,9 +35,9 @@ module Make (K : Irmin.Hash.S) = struct
   module Key = struct
     type t = K.t [@@deriving irmin]
 
-    let hash = Irmin.Type.(unstage (short_hash K.t)) ?seed:None
+    let hash t = Irmin.Type.(unstage (short_hash K.t)) ?seed:None t |> abs
```
Comment (samoht) [inline src/irmin-pack/pack_index.ml:38]: can you replace `abs` by a bitwise operation?

Comment (pascutto) [reply src/irmin-pack/pack_index.ml]: Done

File: src/irmin-pack/pack_index.ml [hunk 2]
```diff
@@ -35,9 +35,10 @@ module Make (K : Irmin.Hash.S) = struct
   module Key = struct
     type t = K.t [@@deriving irmin]
 
-    let hash = Irmin.Type.(unstage (short_hash K.t)) ?seed:None
+    let hash t =
+      Irmin.Type.(unstage (short_hash K.t)) ?seed:None t |> ( land ) 0x7fffffff
```
Comment (samoht) [inline src/irmin-pack/pack_index.ml:39]: can you also inline `|>` ? not sure it's work doing this here :-)

Comment (CraigFe) [inline src/irmin-pack/pack_index.ml:39]: You should hoist `unstage` here to avoid allocations due to generic specialisation in the hot path.

Comment (samoht) [reply src/irmin-pack/pack_index.ml]: Ha yes good catch: @pascutto can you move the `unstage` function out of the closure, so it's allocated only once?

**Comment by CraigFe** [comment]
This seems to have been made redundant by https://github.com/mirage/irmin/pull/1209. Is that correct @pascutto / @samoht?

**Comment by samoht** [comment]
> This seems to have been made redundant by #1209. Is that correct @pascutto / @samoht?

Indeed

**Pull request closed without merging**